cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java

```diff
@@ -137,45 +153,107 @@ public void configure(Map<String, ?> configs) {
+    }
+  }
+
+  protected void maybeUpdateCruiseControlMetricsTopic() {
+    maybeUpdateTopicConfig();
+    maybeIncreaseTopicPartitionCount();
+  }
+
+  protected void maybeUpdateTopicConfig() {
+    try {
+      // Retrieve topic config to check and update.
+      ConfigResource topicResource = new ConfigResource(ConfigResource.Type.TOPIC, _cruiseControlMetricsTopic);
+      DescribeConfigsResult describeConfigsResult = _adminClient.describeConfigs(Collections.singleton(topicResource));
+      Config topicConfig = describeConfigsResult.values().get(topicResource).get();
+      Set<AlterConfigOp> configsToBeSet = new HashSet<>(2);
+      maybeUpdateConfig(configsToBeSet,
+                        LogConfig.RetentionMsProp(),
+                        _newTopic.configs().get(LogConfig.RetentionMsProp()),
+                        topicConfig);
+      maybeUpdateConfig(configsToBeSet,
+                        LogConfig.CleanupPolicyProp(),
+                        _newTopic.configs().get(LogConfig.CleanupPolicyProp()),
+                        topicConfig);
+      if (!configsToBeSet.isEmpty()) {
+        AlterConfigsResult alterConfigsResult = _adminClient.incrementalAlterConfigs(Collections.singletonMap(topicResource, configsToBeSet));
+        alterConfigsResult.values().get(topicResource).get();
       }
+    } catch (InterruptedException | ExecutionException e) {
+      LOG.warn("Unable to update config of Cruise Cruise Control metrics topic {}", _cruiseControlMetricsTopic, e);
+    }
+  }
+
+  protected void maybeIncreaseTopicPartitionCount() {
+    try {
+      // Retrieve topic partition count to check and update.
+      TopicDescription topicDescription =
+          _adminClient.describeTopics(Collections.singletonList(_cruiseControlMetricsTopic)).values()
+                      .get(_cruiseControlMetricsTopic).get();
+      if (topicDescription.partitions().size() < _newTopic.numPartitions()) {
+        _adminClient.createPartitions(Collections.singletonMap(_cruiseControlMetricsTopic,
+                                                               NewPartitions.increaseTo(_newTopic.numPartitions())));
+      }
+    } catch (InterruptedException | ExecutionException e) {
+      LOG.warn("Unable to increase Cruise Cruise Control metrics topic {} partition number to {}",
+               _cruiseControlMetricsTopic, _newTopic.replicationFactor(), e);
     }
   }
 
   @Override
   public void run() {
     LOG.info("Starting Cruise Control metrics reporter with reporting interval of {} ms.", _reportingIntervalMs);
-
     if (_newTopic != null && _adminClient != null) {
-      LOG.info("Try to create Cruise Control metrics topic if not exist");
-      createCruiseControlMetricsTopic();
-      CruiseControlMetricsUtils.closeAdminClientWithTimeout(_adminClient, 5000);
+      try {
+        Collection<TopicListing> topicListings = _adminClient.listTopics().listings().get();
```

Comment:
This call to `listTopics()` introduces an additional network roundtrip in case the topic does not exist.
Wouldn't it be more efficient to attempt to create the topic and handle `TopicExistsException`? If the topic exists, then we can update it.

Maybe more importantly, if the topic is created after the topic list is retrieved here, the call to create topic will fail with `TopicExistsException` and the code won't be able to handle it.